File: solutions/python/backtracking/n-queens-ii.py
# LeetCode 52 — N-Queens II (Hard)
# Category: Backtracking · Approach: Backtracking
# Time: O(n!) | Space: O(n)
# Source: https://leetcode.com/problems/n-queens-ii/

def total_n_queens(n: int) -> int:
    cols: set[int] = set()
    diag1: set[int] = set()
    diag2: set[int] = set()
    count = 0

    def bt(row: int) -> None:
        nonlocal count
        if row == n:
            count += 1
            return
        for c in range(n):
            if c in cols or (row - c) in diag1 or (row + c) in diag2:
                continue
            cols.add(c)
            diag1.add(row - c)
            diag2.add(row + c)
            bt(row + 1)
            cols.remove(c)
            diag1.remove(row - c)
            diag2.remove(row + c)

    bt(0)
    return count

File: solutions/typescript/backtracking/n-queens-ii.ts
// LeetCode 52 — N-Queens II (Hard)
// Category: Backtracking · Approach: Backtracking
// Time: O(n!) | Space: O(n)
// Source: https://leetcode.com/problems/n-queens-ii/

function totalNQueens(n: number): number {
  const cols = new Set<number>();
  const diag1 = new Set<number>();
  const diag2 = new Set<number>();
  let count = 0;
  function bt(row: number): void {
    if (row === n) {
      count++;
      return;
    }
    for (let c = 0; c < n; c++) {
      if (cols.has(c) || diag1.has(row - c) || diag2.has(row + c)) continue;
      cols.add(c);
      diag1.add(row - c);
      diag2.add(row + c);
      bt(row + 1);
      cols.delete(c);
      diag1.delete(row - c);
      diag2.delete(row + c);
    }
  }
  bt(0);
  return count;
}
